### PR TITLE
Adding access cache to JUser

### PR DESCRIPTION
**Pull request by Hackwar**
This adds caching to the authorise method of JUser. Instead of doing the whole check each time JUser::authorise() is called, the checks are done once per session and that is it. This is quite a performance improvement.

To prevent the session from being overrun by the access cache, the number of access results to cache is limited to 250 entries. That should be equivalent to around 10-14kb of data.

I did not make scientific tests with this, just some manual comparisons and I did get positive results in terms of performance gain.

There is a potential issue when an extension creates its own login/logout code that does not use JApplicationCMS::login/logout and which does not correctly reset the JUser object. In that case however, that extension has quite large security issues anyway. 

Anyway, I made the caching variable private, so that we can remove it again in case it turns out to be a failure. There is no backwards compatibility issue there.

### CODE REVIEW

File: libraries/joomla/user/user.php [hunk 1]
```diff
@@ -390,7 +398,26 @@ public function authorise($action, $assetname = null)
...
+		{
+			return true;
+		}
+		else
+		{
+			$key = md5($this->id . '|' . $action . '|' . $assetname);
+
+			if (!array_key_exists($key, $this->accessCache))
+			{
+				if (count($this->accessCache) > 250)
```
Comment (Bakual) [inline libraries/joomla/user/user.php]: Can you add some comments here so we know in a year why we check the count? I'm sure nobdoy will remember it by then.

**Comment by roland-d** [comment]
@Hackwar 
What are the test instructions to make sure this is properly tested?

If I am not mistaken this requires a user to logout and login again to get the correct permissions once it has been changed. How does this relate to the PR #8805? Does this have any effect on this PR?

**Comment by conconnl** [comment]
@Hackwar can we have test instructions and can you answer Roland his question.
We have a PBF now and want to be able to test this.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/6053">issues.joomla.org/joomla-cms/6053</a>.</sub>

**Comment by mbabker** [comment]
With the change in how the `JUser` object is serialized to the session, is this even still valid?  At best it would cache the lookups for a single request cycle based on current staging.

**Comment by tomartailored** [comment]
can we have test instructions<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/6053">issues.joomla.org/joomla-cms/6053</a>.</sub>

**Comment by creativeprogramming** [comment]
They are not interested in performance, just code poetry. ;)

**Comment by mbabker** [comment]
> They are not interested in performance improvements, just 'code poetry'. ;)

Sorry we don't intend to rewrite Joomla as a non-PHP application to follow your scoping and performance preferences 😉

**Comment by Hackwar** [comment]
This PR was meant for when the serialized user object was stored in the session. Since that has changed, I have no idea about the implications that this has on the usefullness of this PR. A cache for the access data still seems to be usefull, but as @mbabker wrote, in the current situation its value is greatly diminished.

**Comment by roland-d** [comment]
@Hackwar Shall we go ahead and close this then? I mean, is it worth the time and effort to pursue this further if the value is greatly diminished?

**Comment by roland-d** [comment]
@hackwar Closing this as the usability has been greatly diminished. If you wish to pursue this further feel free to re-open the pull request. Thank you for your contribution.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/6053">issues.joomla.org/joomla-cms/6053</a>.</sub>

**Pull request closed without merging**